src/main/scala/mesosphere/marathon/state/Group.scala

```diff
@@ -272,23 +285,63 @@ object Group extends StrictLogging {
-      // Only top-level groups are allowed to set the enforce role parameter.
-      if (!groupId.isTopLevel && group.enforceRole.contains(true)) {
-        Failure(Set(RuleViolation(group.enforceRole, s"enforceRole can only be set for top-level groups, and ${groupId} is not top-level", Path(Generic("enforceRole")))))
+    def enforceRoleNotDefined(thisGroup: GroupUpdate, groupId: AbsolutePathId): Result = {
+      if (thisGroup.enforceRole.contains(true)) {
+        Failure(Set(RuleViolation(thisGroup.enforceRole, s"enforceRole can only be set for top-level groups, and $groupId is not top-level", Path(Generic("enforceRole")))))
       } else {
         Success
       }
     }
+
+    override def visit(thisGroup: raml.GroupUpdate): Result = groupUpdateValidator(thisGroup)
+
+    override def appVisitor(): AppVisitor[Result] = AppValidationVisitor()
+
+    override def childGroupVisitor(): GroupValidationVisitor = this
+  }
+
+  case class TopLevelGroupValidationVisitor(override val originalRootGroup: RootGroup) extends GroupValidationVisitor {
+
+    override def childGroupVisitor(): GroupValidationVisitor = LowerLevelGroupValidationVisitor(originalRootGroup)
+  }
+
+  case class RootGroupValidationVisitor(override val originalRootGroup: RootGroup) extends GroupValidationVisitor {
+
+    override def visit(thisGroup: GroupUpdate): Result = super.visit(thisGroup).and(enforceRoleNotDefined(thisGroup, PathId.root))
+
+    override def childGroupVisitor(): GroupValidationVisitor = TopLevelGroupValidationVisitor(originalRootGroup)
+  }
+
+  case class LowerLevelGroupValidationVisitor(override val originalRootGroup: RootGroup) extends GroupValidationVisitor {
+
+    override def visit(thisGroup: GroupUpdate): Result = super.visit(thisGroup).and(enforceRoleNotDefined(thisGroup, PathId.root)) // TODO: pass group path
+
+    override def childGroupVisitor(): GroupValidationVisitor = this
+  }
+
+  case class AppValidationVisitor() extends AppVisitor[Result] {
+    override def visit(app: raml.App, groupId: AbsolutePathId): Result = AppValidation.validNestedApp(groupId)(app)
+  }
+
+  def dispatch(base: AbsolutePathId, update: raml.GroupUpdate, visitor: GroupValidationVisitor, originalRootGroup: RootGroup): Result = {
+    val updateResult = visitor.visit(update)
+
+    // Visit each child group.
+    val childGroupVisitor = visitor.childGroupVisitor()
+    val childrenResult = update.groups.fold[Result](Success)(groups => groups.foldLeft[Result](Success) { (acc, childGroup) =>
+      val absoluteChildGroupPath = PathId(childGroup.id.get).canonicalPath(base)
+      acc.and(dispatch(absoluteChildGroupPath, childGroup, childGroupVisitor, originalRootGroup))
+    })
+
+    // Visit each app
+    val appVisitor = visitor.appVisitor()
+    val appsResult = update.apps.fold[Result](Success)(apps => apps.foldLeft[Result](Success) { (acc, app) =>
+      acc.and(appVisitor.visit(app, base))
```

Comment:
This would be a little simpler with Scalacitc [Or and Every](http://www.scalactic.org/user_guide/OrAndEvery) because the visitor could return `raml.App Or Error` and could stream the app on to the conversion.